### DISCUSSION TITLE
Syncing the states of the Map and the UI

### DISCUSSION THREAD

**Discussion by TangoYankee** [NYCPlanning/ae-zoning-map-poc#26] [Ideas]
# Description

The user interface and the map often display complementary information. Also, updates to one often lead to behavior and state changes in the other. For example, searching for a BBL in the UI may cause the map to zoom to and highlighting the selected BBL. The structure of applications generally complicates the ability to sync these two components. This is because the UI and Map are generally two sibling components with generally their own complex structures. 

Applications with this structure have historically leveraged tools like Redux to manage a global state. However, later versions of React have implemented two hooks that can combine to achieve the same results without relying on a third party tool. This discussion will touch on three general topics. First, it will discuss the structure of Redux. Next, it will discuss the how the same results can be achieved with react hooks, giving specific examples. Finally, it will discuss the implications for Server-Side rendering tools like Next.js.

These general topics will be their own top-level comments, in an attempt to give them their own threads within the discussion.

**Comment by TangoYankee** [comment]
# Redux
A general overview of Redux.

## Philosophy
Before discussing what Redux does, it is important to understand its design strategy. The developers provide a more thorough overview in the [Thinking in redux](https://redux.js.org/understanding/thinking-in-redux/motivation) section of the documentation. Their explanations rely on a familiarity with `pure functions` and `side effects`. [Free code camp has a short primer](https://www.freecodecamp.org/news/what-is-a-pure-function-in-javascript-acb887375dfe/) on these ideas (even highlighting the use of deep cloning v spreading).

In short, redux makes it easier to reason about applications by centralizing the application state and preventing direct mutations of the state. Updates are strictly limited to centralized updater that publishes new copies of the state. This prevents decentralized functions applying conflicting updates that create unpredictable and generally undesirable outcomes.  

## Design
Redux has three key components: The Store, the reducer, and the actions.

### The store
Global application state is held in a store. The store will start with initial values. When updates need to be made to the store, a new copy is made. Modifications are made to that copy. The copy is then published as the new store state. Throughout the application, different components will listen for updates to the store. When a new copy of the store is published, the application components will check whether the specific data they are listening to has changed. If it has changed, then the listening component will update itself accordingly. 

### The reducer
Because the store affects components across a large scope, it needs a method to organize and control how inputs from across its scope are consolidated into a meaningful update. This is the job of a reducer. Requests are sent to the reducer. The reducer then makes a copy of the store, applies the appropriate update based on the request, and then publishes a new copy of the store.

### Actions
The requests made to the reducer are called 'actions'. They are akin to 'events' from the DOM. They typically contain two pieces of information: 1) the type of action that needs to be applied and 2) The minimum amount of data needed to apply that action. Application component will `dispatch` an `action` to the reducer when it requests an update to the state. The reducer can then consolidate the actions from several components into a single meaningful update.

**Reply by TylerMatteo** [reply]
I would be remiss if I didn't immediately point you towards some docs for [TanStack Query](https://tanstack.com/query/latest/docs/react/overview), the library we've been using in this repo thus far. As that page explains, this library is meant to manage _asynchronous_ state. I like libraries like this because they eliminate a lot of the boilerplate necessary to properly manage async state with something like Redux or `useReducer`. And, because of their limited focus, they can be more "opinionated" in their APIs. [This page](https://tanstack.com/query/latest/docs/react/guides/does-this-replace-client-state) explains the relationship between TanStack Query and things like Redux, Zustand, etc. This bit is especially relevant:
```
For a vast majority of applications, the truly globally accessible client state that is left over after migrating all of your async code to TanStack Query is usually very tiny.

There are still some circumstances where an application might indeed have a massive amount of synchronous client-only state (like a visual designer or music production application), in which case, you will probably still want a client state manager. In this situation it's important to note that TanStack Query is not a replacement for local/client state management. However, you can use TanStack Query alongside most client state managers with zero issues.
```

I'll be the first to admit that most of our apps will end up having more client-side state than your average bear, but I still think leveraging a library like this, potentially side-by-side with a client side state manager, is something to strongly consider.

**Reply by TangoYankee** [reply]
I think we're mostly in agreement on managing state. Tan Stack looks like a good complement to the use of client state outlined below. The description of client state below conveniently omits the cases with network calls. In redux, these cases would be handled with thunks- actions that have nested functions that make async calls. For our applications, I agree that using a tool like Tan Stack should take on this role of network calls. This would simplify greatly the custom client side state we would need to manage- things like active layer and feature ids.

However, we should still consider how to structure this client state. I humbly submit the description below for consideration.

**Comment by TangoYankee** [comment]
# React emulation of Redux
Redux comes with many benefits. However, it also comes with overhead. We likely can rely instead on a combination of two react hooks to achieve similar results. These hooks are `context` and `reducer`. 

## Hooks
### Context
The [`useContext` hook](https://react.dev/reference/react/useContext) can be used to emulate redux's `store`. It was originally developed to [alleviate prop drilling](https://react.dev/learn/passing-data-deeply-with-context). It achieved this intended goal by creating an object which is accessible to any components within its scope or 'context'. When this object is updated, the components within its scope check for updates to its relevant data and update themselves accordingly. This is akin to the behavior of the redux store. And by placing context at a global or near-global scope, it can emulate the store.

### Reducer
Another key component of redux is the ability to update its store in a centralized space known a `reducer` that receives `actions`. The React [`useReducer`](https://react.dev/reference/react/useReducer) hook provides a reducer, a dispatcher, and a state. Though, these utilities need to be placed in context to make them available across several components. 

## Example
Let's make the use of these hooks more concrete by providing specific examples. These examples reference my [personal portfolio](https://github.com/TangoYankee/cv), which uses these techniques to manage shared state. Rather copying the code directly, the line numbers will be referenced and hyperlinked, to save space.

### Initial state
Applications need to start with an initial state. This is achieved with a typed POJO. The global state will generally perform best when 1) Only the minimum amount of data needed at a shared level is stored there and 2) The state is avoids nested objects where possible, so that diffing is kept straightforward.

[The type definition in types.ts#L14](https://github.com/TangoYankee/cv/blob/2d12cee45a79f20c13c91476743437d5f497d33c/src/app/state/types.ts#L14) and the 

[The initial state in reducer.ts#L5](https://github.com/TangoYankee/cv/blob/2d12cee45a79f20c13c91476743437d5f497d33c/src/app/state/reducer.ts#L5)

### Context defintion
Context is defined on [context#L5](https://github.com/TangoYankee/cv/blob/2d12cee45a79f20c13c91476743437d5f497d33c/src/app/state/context.ts#L5) with a `state` (in this case `geoState`) and a `dispatcher` (in this case `geoActionsDispatch`). This mirrors the generic structure of the reducer, with a [state and a dispatch](https://react.dev/reference/react/useReducer). (The context is essentially a way to make our reducer available across several components)

## Actions defintion
There are several components related to actions that are outlined within the [types.ts](https://github.com/TangoYankee/cv/blob/main/src/app/state/types.ts)

Each individual action is an object with a `type` and a `payload`. The action `type` is literally a string describing what should be done when receiving the action. This is because the receiving reducer sorts the actions by using a `switch` statement that checks the action `type`. The `GeoAction` on [L22](https://github.com/TangoYankee/cv/blob/2d12cee45a79f20c13c91476743437d5f497d33c/src/app/state/types.ts#L22) is a union because it is passed to the reducer switch statement ([reducer.ts#L13](https://github.com/TangoYankee/cv/blob/2d12cee45a79f20c13c91476743437d5f497d33c/src/app/state/reducer.ts#L13)). The instance of the GeoAction depends on which switch case it matches, hence the union type. The actual values of the strings is defined on [L4](https://github.com/TangoYankee/cv/blob/2d12cee45a79f20c13c91476743437d5f497d33c/src/app/state/types.ts#L4) with the GEO_ACTION_TYPE enum.

The `payload` of the action object contains the minimum data required to carry out the action. Not all actions will have a payload. Actions like 'toggleBlank' would only need to specify that they toggle 'blank'. The reducer will know to look at the current value in state and then toggle it to the complementary value. 

The GeoActionsDispatch on [L36](https://github.com/TangoYankee/cv/blob/2d12cee45a79f20c13c91476743437d5f497d33c/src/app/state/types.ts#L36) is the final piece of the puzzle. It translates the action objects into dispatch functions. The dispatch functions will have the camelCase version of the SCREAMING_SNAKE_CASE action name. Its function parameter will carry the payload of the action. It returns void because it is the reducer that actually returns a new state based on the content of the dispatched action.

So far, we have covered the type defintions of the actions. We can see the actual implementation in [the actions.ts](https://github.com/TangoYankee/cv/blob/main/src/app/state/actions.ts).  We can see on [L6](https://github.com/TangoYankee/cv/blob/2d12cee45a79f20c13c91476743437d5f497d33c/src/app/state/actions.ts#L6) that geoActions is a function with `dispatch` as a parameter and it returns an object of actions. This is because of how the `useReducer` is used with `context`. We want to define the reducer and its associated dispatch once, define the actions with this dispatcher, and then pass these actions to child components for them to dispatch when they need to. When we define geoActions, we create a closure where each action is defined with the same dispatcher. 

We are also defining and maintaining each action in a central location. This makes it easier to reuse the actions among several components and update them all when we need to change features or fix bugs.  

## Reducer definition
The structure of the reducer is defined in [reducer.ts](https://github.com/TangoYankee/cv/blob/main/src/app/state/reducer.ts). It is a function that accepts the current state and the called action. It will match the action.type using a switch statement. Within each switch case, it will create copy of the state and apply the appropriate transformations. It is the only function that is aware of the latest state and all the operations that are being requested of it. Consequently, any logic associated with the update should be held within the reducer. Once the reducer is finished with transforming its copy of the state, it will return the new state object. 

## Providing the reducer
With the components defined, we can bring them together into a complete system. In [main.tsx](https://github.com/TangoYankee/cv/blob/main/src/app/Main.tsx) we define the reducer and place it in a context. [L14](https://github.com/TangoYankee/cv/blob/2d12cee45a79f20c13c91476743437d5f497d33c/src/app/Main.tsx#L14) initializes a reducer with our predefined `geoReducer` and `initialGeoState`. This returns a state (`geoState`) and a dispatch (`geoDispatch`) that we want to provide in context. The `geoState` we can provide directly to the GeoContext provider on [L19](https://github.com/TangoYankee/cv/blob/2d12cee45a79f20c13c91476743437d5f497d33c/src/app/Main.tsx#L19). This makes the state available to child components so they can listen for updates. We could also directly pass the `dispatch` to the context. However, every time a component wanted to dispatch an action, they would need to re-write the action with the dispatch. This is why we first create the `geoActions` closure discussed above. The resulting `geoActionsDispatch` object is then passed to the context. Now, the child components only need to call their action with the appropriate data, rather than needing to rewrite the action.

## Using the reducer
After setting up the reducer, we can see how it is used in two sibling components. We'll specifically focus on the active location id. On [#L20](https://github.com/TangoYankee/cv/blob/2d12cee45a79f20c13c91476743437d5f497d33c/src/app/geoMap.tsx#L20) and #L21of geoMap.tsx we can see the `activeLocationId` state and `updateActiveLocationId` action are destructured from context. On [L31](https://github.com/TangoYankee/cv/blob/2d12cee45a79f20c13c91476743437d5f497d33c/src/app/geoMap.tsx#L31), clicking a location point in the map will trigger an update to the active location state. [L39](https://github.com/TangoYankee/cv/blob/2d12cee45a79f20c13c91476743437d5f497d33c/src/app/geoMap.tsx#L39) listens for updates to this value and changes the map accordingingly.   

No other component uses this exact variable. However, we can imagine for a moment that [container.tsx](https://github.com/TangoYankee/cv/blob/main/src/app/components/resume/container.tsx) uses active 'location' instead of 'position'. In this imagined scenario, the UI container has the same access to the active location and its update action on [L9](https://github.com/TangoYankee/cv/blob/2d12cee45a79f20c13c91476743437d5f497d33c/src/app/components/resume/container.tsx#L9) and L10. In turn, its highlighted state, defined based on 'active location' on [L20](https://github.com/TangoYankee/cv/blob/2d12cee45a79f20c13c91476743437d5f497d33c/src/app/components/resume/container.tsx#L20) will update whenever a point on the map is updated. Correspondingly, the map will update whenever the 'active location' is updated through the container because of the action on [L23](https://github.com/TangoYankee/cv/blob/2d12cee45a79f20c13c91476743437d5f497d33c/src/app/components/resume/container.tsx#L23)

**Reply by dhochbaum-dcp** [reply]
@bmarchena and I implemented this approach yesterday for the [Zoning Districts layer toggle](https://github.com/NYCPlanning/ae-zoning-map-poc/pull/29).  First we implemented the functionality without a complex state management system, and then we [converted](https://github.com/NYCPlanning/ae-zoning-map-poc/pull/29/commits/b86587bb40603b855c7064dc3c84fee2bad856ea) it to this approach.  

My first question is 
The Layers and Filters menu for the app has 4 hardcoded toggles and then the API call to the Zoning Districts endpoint results in another 3 toggles and 21 checkboxes, and then there is a single integer for "Selected".  Looking ahead to how we would implement the rest of the menu, I'm guessing we would need to use the Tanstack hook to fetch the info from the API and then have some sort of function that takes that and puts it into the client side state manager.  Is that correct?

My first thought is that it's a lot of code to create each single boolean.  It's also notable that all of that code could be written just from the same information you get from a single line of code it replaced:
```
const [zoningDistrictVisibility, setZoningDistrictVisibility] = useState<boolean>(false);
```

My second thought is that it is an enormous amount of complexity to create a single boolean.  This process means that just to add a boolean requires editing an additional 3 files.  I think it simply adds too much complexity to our workflow.  If we immediately implemented this across our entire codebase, and we were working with this every day, the amount of complexity would be more than manageable.  But we have to accept that this will never be implemented across our entire codebase and we won't be working with this every day.  I got this laptop two weeks ago and immediately downloaded 26 of our repos that I expect I will need.  There will be periods of multiple months where we will not work on anything which has this system implemented.  I can't speak for everyone, but personally, that means that I'll have to go back and spend all that time re-learning what I learned yesterday.  I don't think that something this complex is appropriate for a portfolio as large and varied as ours.

Is there not a library which simplifies this and does it all in the background?  It feels like there should be something that allows you to just import it and then do something like:
```
const [id, setId] = addToStore({id, type, defaultValue, reducerFunction});
```

**Reply by bmarchena** [reply]
I second David's thoughts on the complexity of the state management implementation. My feeling is the necessary overhead for even smaller features would mean significantly slowing down our workflow and the addition of more code inevitably means more potential for bugs and errors. For example, just yesterday we encountered a bug which crashed the app silently due to a minor typo in one of the boilerplate files.

**Reply by TangoYankee** [reply]
>  I'm guessing we would need to use the Tanstack hook to fetch the info from the API and then have some sort of function that takes that and puts it into the client side state manager. 

We would need to do some validation of the workflow. But yes, the client state would be synced with the async requests by dispatching an action on updates.

**Reply by TangoYankee** [reply]
I looked through the refactor and noticed two things. First, it still relies on component composition and shallow component trees, despite the availability of a context. Second, it doesn't take advantage of the reducer to simplify the action to toggle visibility.

The `zoningDistrictLayer` component is still using component composition, defining itself within the `App.tsx` component where it still has access to the state. This keeps the `App.tsx` component large. This isn't inherently bad. But, now we have a "complex" `App.tsx` AND a "complex" `context`. If we're going to make a context, let's take advantage of it by moving the zoningDistrictLayer into its own file, thereby simplifying the `App.tsx` component.

The `layersFilter` component also relies on a shallow component tree of only two components (`App.tsx -> LayersFilter.tsx`). The `LayersFilters` component itself remains "complex", defining `AccordionItem`, `AccordionButton`, and `AccordionPanel` components within it. With this structure, the `toggleZoningDistrict` only needs to be sent down one layer. However, the toggle isn't actually used until the `AccordianPanel` component. If we wanted to simplify the 'LayerFilters' component by breaking up these sub-components into their own files, we would then need to 'drill' the toggle function down to the actual component that uses it. This is where context comes in, passing data directly to the simplified components. Again, it's fine to have "complex" components OR simple components served by a context. However, mixing the patterns makes for double "complexity".

Finally, the `updateZoningDistrictVisibilty` passes a superfluous data parameter. The reducer already knows the state of the toggle. Passing a parameter adds no new information.

As described in the Actions section of the above description:
> Not all actions will have a payload. Actions like 'toggleBlank' would only need to specify that they toggle 'blank'. The reducer will know to look at the current value in state and then toggle it to the complementary value.

The function can be simplified by removing this parameter from both its definition and use.

action:
```
export const mapActions = (dispatch: Dispatch<MapAction>) => {
  return {
    toggleZoningDistrictsVisibility: () =>
      dispatch({
        type: MAP_ACTION_TYPE.TOGGLE_IS_ZONING_DISTRICT_LAYER_VISIBLE
      }),
  };
};
```

reducer:
```
export function mapReducer(state: MapState, action: MapAction): MapState {
  switch (action.type) {
    case MAP_ACTION_TYPE.TOGGLE_IS_ZONING_DISTRICT_LAYER_VISIBLE: {
      return {
        ...state,
        isZoningDistrictLayerVisibility: !state.isZoningDistrictLayerVisible,
      };
    }
    default:
      return state;
  }
}
```

LayerFilters use case:
```
                <VStack
                  width={20}
                  justify="center"
                  alignSelf={"flex-start"}
                  onClick={toggleIsZoningDistrictLayerVisibile}
                >
```

**Reply by dhochbaum-dcp** [reply]
The placement of zoningDistrictsLayer in App.tsx was not intended to be permanent.  I had a discussion with Tyler where we ascertained the stubs he provided would not work, and I was instructed to put it in App.tsx for now with plans to refactor later.  Let's table the discussion of component composition for the time being.

I also understand that updateZoningDistrictVisibility has a superfluous data parameter.  We left that in because (a) I didn't feel like bothering to figure out how to take it out when it was working, and (b) we have not discussed whether the store will track the state of the toggles and checkboxes, or whether it will track the visibility state of the map layers.  That decision will have implications for how we want to implement functions for the individual zoning layer filters, and I did not want to make an assumption as to whether or not we wanted to keep consistent all of our implementations for managing booleans.

All that said, I don't think you really addressed our concerns about the complexity of the process.  Do you have an example of some code which creates things dynamically like we will need to manage the individual zoning layer filters which will be asynchronous?

**Reply by TangoYankee** [reply]
> I don't think you really addressed our concerns about the complexity of the process.

I tried my best to understand and address your concerns, even briefly giving concerns over asynchronous calls [their own comment](https://github.com/NYCPlanning/ae-zoning-map-poc/discussions/26#discussioncomment-7721570). It seems I fell short in addressing them.

Before I give another explanation that does not address your underlying concerns, I want to make sure I have a clear picture of your understanding. Could you explain your understanding of what role this client-state manager is meant to fill? Please include your understanding of how it differs from the server-state manager.

**Reply by dhochbaum-dcp** [reply]
My understanding is that a server-state manager is meant to manage data fetched asynchronously from a database/API, that data not having manipulated locally; and that a client-state manager provides a store for data and variables created on the client side.  What I don't understand is how to take the words in your response regarding asynchronous calls and translate it into code.  I've demonstrated in the code that I understand how to take a layer that we know exists (in this case, the Zoning Districts layer that is toggled by clicking the image) and create a variable in the state and a reducer to update it.  But we are not pre-supposing to know that the contents of the Zoning Districts accordion is those 3 toggles and 21 checkboxes because we are pulling that information from the API, and so we can't just add 24 more booleans to store their state like we did previously.  What I don't understand is how to take that data from the API which tells me what toggles and checkboxes will exist, and convert it into types/actions/reducers.  I simply don't understand how to manage the variables we don't know until we fetch them, and 
> the client state would be synced with the async requests by dispatching an action on updates. 

either doesn't address this, or doesn't explain in a way that I understand what to do.

My concern is that Bryan and I paired, followed your thorough documentation and relatively simple code examples, and working together it took us a non-trivial amount of time (maybe an hour?) to understand and implement.  At some point I am going to stop working on this, work on other things, and then come back having forgotten everything I learned.  At that point, if I want to do something as simple as add another boolean, I'm going to have to go back and re-learn what I need to know to even know that I need to make updates to another three files in addition to the one or two files in which I'm actually using the variable.  So that's a significant amount of time to learn how to edit multiple files, when all of that same information can be expressed as simply as 
```
Create a new variable that's named [id] and is a [type] type, with [defaultValue] as the default and [function] to update it.
```
Is there not a solution that implements this process in a way which someone completely unfamiliar with the process would understand without having to dig into documentation?  Is there not a way that someone can open a file and see "hey, here's all the variables in the store, it looks like to create a new one I just need to provide an id, a type, a default, and a reducer; this syntax makes it really clear how to do that"?

Bryan agrees with me that this solution is too complex.  I'd like to remind you that in addition to maintaining the various Ember and React apps and APIs in our portfolio, I also:
- Manage all aspects of our web analytics throughout the agency across multiple platforms
- Maintain the Matomo server repo
- Maintain a decrepit Symfony website builder application
- Work on creating an alternate to that website in Alchemer using LUA and vanilla JS to do things that platform was never intended to do
- Have to fix any issues with another site build with Drupal
- Write and update VB scripts required to create the Excel downloads for EDDE
- Own the agency's pair of Squarespace sites
- Learned Teamsite, and will need to have a working understanding of whatever system the new main site for the agency is using

So it's just more complex than I'd like it to be, given the myriad of languages/platforms/frameworks I need to be switching between on a regular basis, on top of the fact that I am very, _very_ stupid.  I would prefer if there was something that could accomplish this in a way that requires that I remember a lot less and figure out what to do a lot quicker.

**Reply by dhochbaum-dcp** [reply]
Have we considered using Zustand?  I was able to do the [same conversion](https://github.com/NYCPlanning/ae-zoning-map-poc/pull/30/commits/93d419335276eff84aa88d99e07aed3702b2dc0f) using Zustand in 15 minutes using only the code shown on [this demo](https://zustand-demo.pmnd.rs/) as a reference.

**Reply by horatiorosa** [reply]
I _somewhat_ am able to follow this convo and took a look at PR's #29  and #30. 

IMHO (with much emphasis on the H), I find that the [Zustand](https://github.com/pmndrs/zustand) PR is easier and less confusing to read / follow.

**Reply by TangoYankee** [reply]
Zustand is interesting and I wouldn't rule it out to reduce boilerplate. Also, I have been working to create an example of how to dynamically hold the visibility of each layer in state, without needing to hand-write each boolean.

**Reply by TangoYankee** [reply]
Three problems have been conflated in this discussion and I want to refocus.
1) The question of how to pass client state across client components.
    - This is intended focus of this discussion. 
    - The prevailing opinion is that the boilerplate of the reducer is burdensome. A tool like [Zustand](https://github.com/pmndrs/zustand) could remove the boilerplate. We should continue to discuss here whether Zustand is the best solution, compared to other client side state management tools or techniques, like sticking with component composition or moving to another client state management tool.
2) The question of how to create a data structure that can handle the visibility information of layers that are dynamically loaded.
   - This problem is distinct from how to make the data available across the application. I created discussion #31 to brainstorm how to solve that problem.
3) The question of moving between various tech stacks
    - This concern should be raised directly with leadership. The problem of managing client state remains, regardless of whether other applications need to be maintained.

**Reply by dhochbaum-dcp** [reply]
2 - Thank you, that was extremely helpful
3 - I mentioned that only with the intent to discuss how to manage client state, not to expand the scope of the discussion. I intended only to emphasize why I feel that the simplicity of the client state management solution is so important.
1 - Great, so it seems as if we have established so far that Zustand > manual reducer boilerplate.  In my searching, it appeared to be the most simple of the state management solutions.  It:
 - Consolidates everything down to one file with two objects
 - Makes adding a new variable two lines for each object, one for the variable and one for the function
 - Uses common syntax and is extremely easy to parse and understand

Does anyone have any concerns about moving forward with Zustand or have another tool they would like to be considered?  @TylerMatteo you haven't given your two cents yet.  If there are no concerns/alternates to consider, I will begin to work on consolidating the concepts of the following three branches into one:
- Tim's implementation of managing state based on [dynamic layer](https://github.com/NYCPlanning/ae-zoning-map-poc/discussions/31#:~:text=Use%20array%20of%20active%20layers) returns
- The [implementation](https://github.com/NYCPlanning/ae-zoning-map-poc/pull/30/commits/93d419335276eff84aa88d99e07aed3702b2dc0f) of Zustand
- Changes I made to the [user interface](https://github.com/NYCPlanning/ae-zoning-map-poc/pull/29/files) after implementing the React emulation of Redux

**Reply by TylerMatteo** [reply]
@dhochbaum-dcp I'm on board with trying out Zustand first. How we manage client state is certainly something we'll need to find an answer for eventually, and Zustand is as good a place to start as any. I could definitely see us needing to implement the reducer pattern (either through React's built in hooks or something like Redux) eventually as app complexity grows. Zustand is bound to be easier to read and contain less boilerplate for a relatively simple app but, at some point, reducers actually make code _easier_ to follow.

On the topic of "dynamic layers", are you planning to change your layer visibility booleans in #30 to instead use an array of layer IDs? I definitely think an approach like that is better in the long run, but I can I live with a couple bools for this POC, given that we know it's only going to be tax lots and zoning districts.

**Reply by dhochbaum-dcp** [reply]
So there are two ways to do this - to hold in the store the state of the input elements, or to hold the state of the layer elements.  (You could probably do both or a combination of part of each, but those seem like bad ideas).  I decided to go with the input elements, because it made more sense to me.  

Moving forward, I was planning on using top level booleans + layer group toggle array + layer ids checkbox array, which to me more intuitively reflects the structure in the menus.  It also helps indicate that the impact on lower levels of toggling a higher level is not consistent.  It also produces easy to understand code - a layer will be visible if( topLevelBoolean || (layerGroupArray.includes(groupId) && layerArray.includes(layerId)).

This was the way that made the most sense to me, but I'm open to hearing your thoughts.

**Reply by TylerMatteo** [reply]
Keep in mind, I'm using the term "layer" here to refer to the `MVTLayer` class instances that we're instantiating. So, by that definition, all zoning districts are one "layer". I think a good distinction we should make that can inform how we structure the state is the difference between the following:

* A layer (such as the `MVTLayer` for zoning districts) being visible or not. In the code, this is reflected by the `visible` property on the layer
* A layer being _filtered_ based on some combination of properties of the data for that layer. For us, this is things like filtering based on the ZD class category (e.g. "Residential") or based on the class itself (e.g. "C1"). I'll talk about _where in the code_ this filtering happens below.

For the purposes of this discussion, let's just focus on Zoning Districts - the fact that the Tax Lots section has two top-level switches for "Tax Lot Boundaries" and "Land Use Colors" introduces some complexity that we can come back to once we settle on an approach for ZDs.

## How do we manage layer visibility?
This refers to the layer's `visible` property. I suggest we only change this property based on whether or not the entire layer is "on or off". In our UI, this refers to the button with the rounded corner zoning districts image, and _not_ any of the filter toggles. I think using top level booleans for this is a perfectly ok place to start. Other approaches would be having an array of layer ids, or even some sort of solution that takes advantage of Deck's `layerFilter` [property](https://deck.gl/docs/api-reference/core/deck#layerfilter). Note that this approach means you wouldn't do something like your `if( topLevelBoolean || (layerGroupArray.includes(groupId) && layerArray.includes(layerId))` pseudocode.

## How do we manage filters on the data.
This is where things get a little hairy, but I'll do my best to help. 

> Note that the following advice assumes that the data in the Zoning Districts tileset matches the schema we laid out for DE in [this Issue](https://github.com/NYCPlanning/data-engineering/issues/284). When I log out `f.properties` in `getFillColor`, the data don't appear to match that. We'll sync up with DE about getting that fixed.

We know that Deck.gl "accessor function" such as `getFillColor` convert the data from the tile set into geojson features. Given that, we might expect `feature.properties" for a C1 district to look something like this:
```
{
 id: "some-uuid",
 district: "C1-A",
 commercial: "C1",
 manufacturing: null,
 residential: null
}
```
All of our filter logic will be based on the commercial, manufacturing, and residential properties. "district" will have things like "R6B" and it only mean to be used for rendering the label.

I suggest we stay away from borrowing terms from the existing Zola/layers API implementation such as "layer group". For our purposes, let's just stick to "layer" referring to the Deck.gl `MVTLayer` instance. Therefore, we just have a "zoning districts" layer. 

> Moving forward, I was planning on using top level booleans + layer group toggle array + layer ids checkbox array

Terminology aside, I think you're on the write track here with the two arrays. I could imagine your store looking something like this:
```
export const useStore = create<Store>()((set) => ({
 // other existing state...
  zoningDistrictsFilters: {
    class: ["C1", "R2"], // As the user toggles the checkboxes, values are added/removed from this array
    category: ["manufacturing"] // Same here, but for the category Switches
  }
}));
```
Then you could do logic with `.includes`. However, another approach you should consider would be instead having these be objects with key/value pairs. That way, `class: ["C1", "R2"]` becomes:
```
class: {
  C1: true,
  C2: false,
  // and so on
}
```
> Note that in the approach I'm suggesting, the filters state is "flat". That is to say, the state for all of the zd class toggles is under `class`, even though each could be considered a "child" of it's corresponding category.

### Where does the filtering logic happen in the code?
Ok, so if we have state for the various checkboxes and switches, and we know we're only using the Layer `visible` property to turn the _entire layer_ on or off, where do we actually put the _filtering_ logic? There are two approaches I see here:

#### 1. Filtering by showing invisible features
The easier way to implement this is to simply render transparent features by doing the filter logic in functions like `getFillColor`. It's important to note that, as far as Deck is concerned, this isn't really "filtering" anything, it's just rendering a feature with 0 opacity. In this implementation, you would be comparing the features on the data in the tile to the state controlled by the checkboxes and switches

#### 2. Filtering with [DataFilterExtension](https://deck.gl/docs/api-reference/extensions/data-filter-extension)
Deck does offer an extension for doing GPU-based data filtering. This will be more performant because, once a particular polygon is filtered out, functions like `getFillColor` will never run for that polygon at all. It even supports filtering on multiple properties (expressed as nested arrays of value ranges to filter on). However, currently this extension only works with numeric properties:
```
getFilterValue: f => f.properties.timeOfDay,  // in seconds
filterRange: [43200, 46800],  // 12:00 - 13:00
```
In our cases, we want to filter on _categorical_ string properties. The good folks maintaining Deck are [working on](https://github.com/visgl/deck.gl/issues/7827) a categorical filter but it hasn't been released yet.

If we want to use this extension in the meantime, it should theoretically be possible to "coerce" our use cases to numbers via `getFilterValues`. That would look something like:
```
{
  getFilterValue: feature => {
    // Do logic with feature.properties and your store to determine if this 
    // feature should be shown. If it should, return 1, else return 0
  },
  filterRange: [1, 1], 
}
```
But this approach might introduce more janky code than it's worth or may be slow. If thats the case, option 1 would be the fallback.

**Reply by TangoYankee** [reply]
Cross-referencing discussion #32

**Reply by dhochbaum-dcp** [reply]
> Note that the following advice assumes that the data in the Zoning Districts tileset matches the schema we laid out for DE in https://github.com/NYCPlanning/data-engineering/issues/284. When I log out f.properties in getFillColor, the data don't appear to match that. We'll sync up with DE about getting that fixed.

That data is already in the `district` field, so the tiles don't need to be redone.  Tweaking my code to the following should be all we need:
```
    getFilterValue: (f: any) => {   
    /**
    *   Show if either:
    *     allZoningDistrictsVisibility is true, or
    *     The matching category is toggled to true and the the class checkbox is true 
    *     for any matching districts
    */
      return (allZoningDistrictsVisibility ||
             (visibleZoningDistrictCategories.has(f.properties.district[0])) &&
             (visibleZoningDistrictClasses.has(f.properties.district.match(/\w\d*/)[0]) || visibleZoningDistrictClasses.has(f.properties.district.split('/')[1]?.match(/\w\d*/)[0]))) ?
              1 : 0;
    },
    filterRange: [1, 1],
```

**Reply by TylerMatteo** [reply]
I was actually incorrect in that note you quoted. The polygons in the tileset that we're using to display the zoning districts actually _do_ have the "commercial", "residential", and "commercial" properties we need. The difference between the current tileset we're using and the schema laid out in the [issue](https://github.com/NYCPlanning/data-engineering/issues/284) is that those 3 fields will simply be undefined instead of existing with `null` value. That's why my [changes](https://github.com/NYCPlanning/ae-zoning-map-poc/pull/30/commits/4bc55a10bdcf074377bff461de47a288f7a0dde6#diff-26ad4b834941d9b19ebf9db8082bd202aaf72ea0ddea85f5a8a0cb3c729cc6f2R75) to `getFilterValue` loop through the three categories and then use `hasOwnProperty` to see if the property is defined on each polygon.

The tile set also contains data that we use to show the _labels_, in the form of "Point" geometries. Those are missing the "commercial", "residential", and "commercial" properties we need, which is why I put in this [circuit breaker if statement](https://github.com/NYCPlanning/ae-zoning-map-poc/pull/30/commits/4bc55a10bdcf074377bff461de47a288f7a0dde6#diff-26ad4b834941d9b19ebf9db8082bd202aaf72ea0ddea85f5a8a0cb3c729cc6f2R68) we need for now until the tile set is updated.

As for the matter of using string splitting and regex matching, there is one change in my commit that I forgot to point out. I [updated](https://github.com/NYCPlanning/ae-zoning-map-poc/pull/30/commits/4bc55a10bdcf074377bff461de47a288f7a0dde6#diff-1de282060139845e13d4879386bcfb6e62156aef95e04924e231805cdbf70f9cR62) the call to `toggleZoningDistrictCategoryVisibility` to push in the whole category name (so "Manufacturing", "Residential", or "Commercial") and lowercased it so it can be matched against the 3 corresponding properties in the tile set data. That is how I'm able to loop through the visible categories in `getFilterValue` and perform the show/hide logic with `hasOwnProperty`.

I believe the `getFilterValue` function as written in my commits is good to go for now. However, as a rule of thumb, we're trying to avoid code that tries to figure out properties of data points by splitting or parsing apart strings that "encode" data because it's unpredictable and leads to us writing difficult-to-understand code on the frontend. We want to avoid this by building that information we need into the database tables and tile sets we use up front. In this particular case, that's what we're doing with the "commercial", "residential", and "commercial" properties on the data in the tile set.

**Comment by TangoYankee** [comment]
# Global State and Server Side Rendering

When components are styled based on client state, this complicates server side rendering. To maintain server-side rendering, the state needs to be mirrored on the server. This greatly increases complexity. The alternative is to fallback to client side rendering for any component within the context of the state. For our applications, this will be the vast majority of components. Coupled with the fact that our maps can only be client side rendered, it calls into question the value of attempting to use server side rendering at all.

**Reply by TylerMatteo** [reply]
I agree that we shouldn't take it for granted that we need a way to do server side rendering for every app. The approach we should take is "Use server side rendering and static site generation, unless you have a good reason not to". This approach of defaulting to static generation, then to server side rendered, then to totally client side is well documented in places such as this [18F engineering guide page](https://engineering.18f.gov/web-architecture/).

Also on this topic, I encourage you to look at how asynchronous state libraries like Tanstack Query and [SWC](https://swr.vercel.app/) handle SSR and SSG. It usually looks something like [this](https://tanstack.com/query/latest/docs/react/guides/advanced-ssr#prefetching-and-dehydrating-data). You basically make API calls on the server side and "preload" that data into the cache of data that is available on the frontend.

**Reply by TangoYankee** [reply]
I understand the need for having a way to handle async calls. But, I also want to make sure we're engaging with this thread on its own terms- as a client side state manager,
